src/models/release-asset.ts

```diff
@@ -0,0 +1,468 @@
+			return buffer.subarray(0, bytesRead);
+		} finally {
+			await fd.close();
+		}
+	};
+
+	const uploadPart = async (
+		asset: File | string,
+		part: UploadPart,
+		requestedPartSize: number,
+	) => {
+		let body: Blob | ArrayBuffer;
+		const offset = (part.partNumber - 1) * requestedPartSize;
+
+		if (typeof asset === 'string') {
+			const buffer = await readFileChunk(asset, offset, part.chunkSize);
+			body = buffer.buffer.slice(
+				buffer.byteOffset,
+				buffer.byteOffset + buffer.byteLength,
+			);
+		} else {
+			const end = offset + part.chunkSize;
+			body = asset.slice(offset, end);
+		}
+
+		// We directly use fetch API rather than balena-request because
+		// pre-signed S3 URLs contain authentication in query parameters
+		// and must be used exactly as provided without modification (encoding/decoding for balena-request)
+		const res = await fetch(part.url, {
+			method: 'PUT',
+			body,
+		});
+
+		const ETag = res.headers.get('ETag')?.replace(/^"+|"+$/g, '');
+		if (ETag == null || typeof ETag !== 'string') {
+			throw new Error(`Error on the received ETag ${ETag}`);
+		}
+		return {
+			PartNumber: part.partNumber,
+			ETag,
+		};
+	};
+
+	const uploadParts = async (
+		asset: File | string,
+		parts: UploadPart[],
+		uploadParams: Required<Pick<UploadParams, 'chunkSize' | 'parallelUploads'>>,
+	) => {
+		const limit = pLimit(uploadParams.parallelUploads);
+		return await Promise.all(
+			parts.map((part) =>
+				limit(() => uploadPart(asset, part, uploadParams.chunkSize)),
+			),
+		);
+	};
+
+	const uploadMultipartReleaseAsset = async (
+		releaseAssetId: number,
+		asset: File | string,
+		uploadParams: Required<Pick<UploadParams, 'chunkSize' | 'parallelUploads'>>,
+	) => {
+		let metadata;
+		if (typeof asset === 'string') {
+			const size = await getFileSize(asset);
+			metadata = {
+				filename: getFileBasename(asset),
+				content_type: getMimeType(asset),
+				size,
+			};
+		} else {
+			metadata = {
+				filename: asset.name,
+				content_type: asset.type || 'application/octet-stream',
+				size: asset.size,
+			};
+		}
+
+		const beginUploadResponse = (await pine.post({
+			resource: 'release_asset',
+			id: releaseAssetId,
+			action: 'beginUpload',
+			body: {
+				asset: {
+					...metadata,
+					chunk_size: uploadParams.chunkSize,
+				},
+			},
+		})) as ReleaseAssetBeginUpload;
+
+		await pine.post({
+			resource: 'release_asset',
+			id: releaseAssetId,
+			action: 'commitUpload',
+			body: {
+				uuid: beginUploadResponse.asset.uuid,
+				providerCommitData: {
+					Parts: await uploadParts(
+						asset,
+						beginUploadResponse.asset.uploadParts,
+						uploadParams,
+					),
+				},
+			} satisfies CommitData,
+		});
+	};
+
+	const createMultipartReleaseAsset = async (
+		createParams: WriteReleaseAssetParams,
+		uploadParams: Required<Pick<UploadParams, 'chunkSize' | 'parallelUploads'>>,
+	) => {
+		const releaseAssetId = await getOrCreateReleaseAsset(createParams);
```

Comment:
I think that you should be able to do:
```suggestion
		const releaseAssetId = (await pine.getOrCreate({
			resource: 'release_asset',
			id: { release, asset_key },
			options: {
				$select: 'id',
			},
			body: {
				asset: null,
			},
		}))!.id;
```